### ISSUE TITLE
[FEA] Accelerate cupy array creation from DataFrame.values

### ISSUE THREAD

**Issue by bdice** [rapidsai/cudf#16483]
**Is your feature request related to a problem? Please describe.**
Users with large numerical datasets (such as a dataframe with thousands of time-series columns) would like to be able to convert from a cuDF dataframe to a cupy array as quickly as possible. Currently we have a raw loop in Python that does casting and assignment for each column. https://github.com/rapidsai/cudf/blob/a8a367009ff64478d78eb916fc9dc65b77b89aac/python/cudf/cudf/core/frame.py#L458-L463

It should be possible to lower this into libcudf, and use a kernel that calls a batched memcpy from CCCL CUB to copy the same-type input columns into the matrix directly. Some columns may require casting, and that work could be launched in separate streams.

**Describe the solution you'd like**
```cpp
template<typename T>
void table_to_array(cudf::table_view input, cudf::device_span<T> output) {
    CUDF_EXPECTS(std::all_of( ... /* all input columns are convertible to T */, cudf::data_type_error));
    // 1. Get boolean map of which columns already match the output type
    // 2. Call batchmemcpy on all matching columns
    // 3. Use a thrust transform with custom input and output iterators for casting all other types
    //    This is nontrivial but shouldn't be too hard. Some device-side type dispatch, maybe.
}
```

**Describe alternatives you've considered**
~We might be able to use `cudf::contiguous_copy_column_device_views`, but that requires all the types to be the same.~ (I was wrong, this is not what I want.) I think the best performing solution would do casting of any compatible input type to the target type as it copies.

We could also make the API take a `void *` and a `cudf::data_type output_dtype`? I'm not sure. I think it is important for this to have an output parameter and let the data be allocated by cupy with `matrix = cupy.empty(shape=(len(self), ncol), dtype=dtype, order="F")` [like we already do here](https://github.com/rapidsai/cudf/blob/a8a367009ff64478d78eb916fc9dc65b77b89aac/python/cudf/cudf/core/frame.py#L455-L457).

**Comment by bdice** [comment]
> We could also make the API take a `void * output` and a `cudf::data_type output_dtype`?

Yes, this is probably the right way. We ought to do host-side type-dispatch to determine which kernel to call, and device-side type-dispatch to handle casting the various input column types.

**Comment by bdice** [comment]
We might be able to shortcut this and just accelerate the "easy" path where all types are the same, to start out. For that we can just use the same logic as `cudf::contiguous_copy_column_device_values`, and only call this from Python if we know all the types match.

**Comment by mroeschke** [comment]
xref https://github.com/rapidsai/cudf/issues/11648

**Comment by bdice** [comment]
Also xref #12928

**Comment by bdice** [comment]
Sharing some notes from a related offline discussion:

- cuDF table -> cupy array:
  - cannot be a view because columns aren't contiguous
  - this is the slow thing that I want to fix in this issue
- cupy array -> cuDF table:
  - can be a view for each column (if `F` ordered)
  - typically arrays are _not_ `F` ordered unless you made them that way on purpose. In `C`-ordered cases, a transpose is needed.
- cuDF lists column -> cupy array:
  - can be a view (zero-copy, produces `C` ordered output)
  - Snippet: `ser.list.leaves.values.reshape(len(ser), -1)`
- cupy array -> cuDF lists column
  - can be a view (if `C` ordered) but requires offsets construction
  - https://github.com/rapidsai/cudf/issues/18214

**Comment by wence-** [comment]
> cuDF lists column -> cupy array:

>    can be a view (zero-copy, produces C ordered output)
>    Snippet: ser.list.leaves.values.reshape(len(ser), -1)

Only if every row has the same length list.

**Comment by bdice** [comment]
> Only if every row has the same length list.

Yes, sorry. That was assumed in the use case that motivated this discussion earlier and I forgot to say it explicitly.

**Comment by Matt711** [comment]
Some follow-ups after #18450 is merged
- https://github.com/rapidsai/cudf/pull/18450#issuecomment-2807574132
